archive/wg-component-standard/component-config/README.md

```diff
@@ -0,0 +1,977 @@
+        *   We _shouldn't_ ever **fix bad defaults** for existing flag names without bumping the major version of a binary. In reality, we get around this by giving advance warning and technically breaking semantic versioning of the binary.
+        *   We _shouldn't_ ever **remove a flag **without bumping the major version of a binary. In reality we use a [flag deprecation policy](https://kubernetes.io/docs/reference/deprecation-policy/#deprecating-a-flag-or-cli) that allows us to technically break semantic versioning of the binary as long as we give advance warning.
+    *   We incrementally deprecate individual parameters over time, instead of guaranteeing a consistent set of parameters for the life of an API version. This confuses users and results in a less stable API.
+    *   We can't typically deploy flag-based configuration independently from a binary version upgrade, because the compatibility of the interface is so tightly coupled.
+*   Values are often re-configured, which precipitates additional tools to parameterize and write configuration for system-specific process management agents (e.g. systemd). We can eliminate the dependency on parameterization tools if process manager config is static; e.g. the configured command line just needs to reference a file in a fixed location.
+*   Developers inevitably embed structured data in strings and invent one-off parsers to process their flags. This invites bugs.
+*   mikedanese@ outlined more issues in his [document](https://docs.google.com/document/d/1arP4T9Qkp2SovlJZ_y790sBeiWXDO6SG10pZ_UUU-Lc).
+
+Core goals of componentconfig include:
+
+
+
+*   Standardize the configuration approach for all core cluster components.
+*   Enable dynamic configuration deployment mechanisms.
+
+Conveniently, Kubernetes has similar goals:
+
+
+
+*   Standardize the configuration approach for cluster infrastructure.
+*   Enable dynamic deployment mechanisms.
+
+Kubernetes had already paved the way: It has what we need to version our configuration interfaces, decouple configuration changes from binary changes, represent configuration in a structured format, and deploy configuration in a dynamic environment.
+
+
+
+*   Versioning was accomplished via the API machinery's group/version mechanism.
+*   Adhering to the same API guarantees as the core Kubernetes APIs provides a stable configuration surface, and allows us to decouple the configuration interface from releases that support the same API version.
+*   Kubernetes API objects are defined as Go structs, which means we don't have to parse strings to deal with structured config.
+*   Kubernetes has deployment mechanisms (ConfigMap volume source) that work well for pushing new versions of configuration files into production.
+*   There is no requirement to restart a process when you change a file, unlike flags.
+
+**All core Kubernetes components should eventually consume their configuration via versioned configuration files, instead of command-line flags. **
```

Comment:
```suggestion
**All core Kubernetes components should eventually consume their configuration via versioned configuration files, instead of command-line flags.**
```

to maintain emphasis